agent/proxycfg/state.go

```diff
@@ -432,6 +437,54 @@ func (s *state) initWatchesMeshGateway() error {
 	return err
 }
 
+func (s *state) initWatchesIngressGateway() error {
+	// Watch for root changes
+	err := s.cache.Notify(s.ctx, cachetype.ConnectCARootName, &structs.DCSpecificRequest{
+		Datacenter:   s.source.Datacenter,
+		QueryOptions: structs.QueryOptions{Token: s.token},
+		Source:       *s.source,
+	}, rootsWatchID, s.ch)
+	if err != nil {
+		return err
+	}
+
+	// Watch the leaf cert
+	err = s.cache.Notify(s.ctx, cachetype.ConnectCALeafName, &cachetype.ConnectCALeafRequest{
+		Datacenter:     s.source.Datacenter,
+		Token:          s.token,
+		Service:        s.service,
+		EnterpriseMeta: s.proxyID.EnterpriseMeta,
+	}, leafWatchID, s.ch)
+	if err != nil {
+		return err
+	}
+
+	// Watch the service's own ingress-gateway config entry
+	err = s.cache.Notify(s.ctx, cachetype.ConfigEntryName, &structs.ConfigEntryQuery{
+		Datacenter:     s.source.Datacenter,
+		QueryOptions:   structs.QueryOptions{Token: s.token},
+		Kind:           structs.IngressGateway,
+		Name:           s.service,
+		EnterpriseMeta: *structs.WildcardEnterpriseMeta(),
+	}, ingressGatewayConfigWatchID, s.ch)
+	if err != nil {
+		return err
+	}
+
+	// Watch the list of all services in the datacenter.
```

Comment:
Shouldn't this only list the services defined in the config entry?

Reply:
The reason we watch the list of all services is so that we can handle the case where a user defines a wildcard service name (within a namespace).

The current code goes like this:
1. Watch all services in datacenter
2. On update event for `service-list`, construct a map of `namespace -> set of services in namespace`.
3. Call `resetIngressUpstreamsAndDiscoveryWatches`, which will iterate over the defined config entry's Listeners and Services.
4. If a user defined a service with the wildcard specifier, we lookup the namespace defined in the map constructed in step 2, and add every known service in that namespace to the list of upstreams.

So, we need to watch the list of all services, in every namespace, in order to handle dynamic updates of services/config entries if a wildcard entry is specified.

Reply:
This differs from how the current terminating gateway PR [handles the wildcard case](https://github.com/hashicorp/consul/pull/7571/files#diff-2271ca70cf81490eff6dd6060d3f35b5R2395-R2400), where we deal with wildcard entries by creating a new state store table to store the mapping.

We should decide collectively which approach is better and converge. cc @freddygv

Reply:
In conversation with @freddygv, we've decided to converge on a new RPC that returns just the list of services that are configured by the config entry. 👍 

We still have to converge on the underlying implementation in the state store around how we compute those values, but that is easier to fix later, and converging on the RPC is more important now.

Reply:
We've refactored the state store and RPC parts of this to use the same RPC/memdb table that terminating gateways are using.